docs/language/informal/toplevel-inference.md

```diff
@@ -0,0 +1,338 @@
+   the inferred types of the provided keys, and `V` is the least upper bound of
+   the provided values.
+   * If the keys and values do not all have an inferred type, it is an error.
+   * The inference dependencies of the map literal are the collected
+      dependencies of the keys and values.
+* An non-const empty map literal with no type arguments has the type
+  `Map<dynamic, dynamic>`.
+   * No inference dependencies.
+* A const empty map literal with no type arguments has the type `Map<Null,
+  Null>`.
+   * No inference dependencies.
+* A function literal with an expression body for which all parameter types are
+   type annotated and for which the return expression is an IE and has an
+   inferred type, has the corresponding function type.
+   * If the return expression has no inferred type it is an error.
+   * The inference dependencies are the inference dependencies of the return
+   expression.
+   * If the function is marked `async`, then the return type of the function is
+     inferred as `Future<flatten<T>>` where `T` is the inferred type of the
+     return expression and flatten is as described in the spec.
+* An instance creation expression with no omitted type arguments has the
+   obvious type.
+   * No inference dependencies
+* An instance creation expression with no omitted type arguments has the
+   obvious type.
+   * No inference dependencies
+* An `as` expression has the cast to type
+   * No inference dependencies
+* An `is` expression has type boolean
+   * No inference dependencies
+* A simple or qualified identifier referring to a top level function, static
+   variable, field, getter; or a static class variable, static getter or method;
+   or an instance method; has the inferred type of the identifier.
+   * Otherwise, if the identifier has no inferred or annotated type then it is
+   an error.
+   * Note: specifically, references to instance fields and instance getters are
+     disallowed here.
+   * The inference dependency of the identifier is itself if the identifier is
+      a candidate for inference.  Otherwise there are no inference dependencies.
+* A simple identifier denoting a formal parameter which has an annotated type is
+  an immediately evident expression and has the type with which it was
+  annotated.
+   * No inference dependencies.
+* A function (or function expression) invocation with no omitted generic
+   arguments where the applicand is an IE that has an inferred type with a
+   return type of `T`, has type `T` (with any generic arguments substituted in).
+   * If the applicand has no inferred type, it is an error.
+   * The inference dependencies are the inference dependencies of the
+      applicand.
+* A method invocation `o.m(...)` with no omitted type arguments where the
+   receiver is an IE with inferred type `T` such that `T` provides a signature
+   for `m` with return type `S` has type `S` (with any generic arguments
+   substituted in).
```

Comment:
Yes.  In general, the intention is that the inferred toplevel type should be the same as the inferred type in a local context.  I added bullets for binary and unary expressions.